massa-bootstrap/src/server/allow_block_list.rs

```diff
@@ -0,0 +1,114 @@
+use std::{
+    borrow::Cow,
+    collections::HashSet,
+    net::{IpAddr, SocketAddr},
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use massa_logging::massa_trace;
+use parking_lot::RwLock;
+
+use crate::tools::normalize_ip;
+
+/// A wrapper around the allow/block lists that allows efficient sharing between threads
+// TODO: don't clone the path-bufs...
+#[derive(Clone)]
+pub(crate) struct SharedAllowBlockList<'a> {
+    inner: Arc<RwLock<AllowBlockListInner>>,
+    allow_path: Cow<'a, Path>,
+    block_path: Cow<'a, Path>,
+}
+
+impl SharedAllowBlockList<'_> {
+    pub(crate) fn new(allow_path: PathBuf, block_path: PathBuf) -> Result<Self, String> {
+        let (allow_list, block_list) =
+            AllowBlockListInner::load_allow_block_lists(&allow_path, &block_path)?;
+        Ok(Self {
+            inner: Arc::new(RwLock::new(AllowBlockListInner {
+                allow_list,
+                block_list,
+            })),
+            allow_path: Cow::from(allow_path),
+            block_path: Cow::from(block_path),
+        })
+    }
+
+    /// Checks if the allow/block list is up to date with a read-lock
+    /// Creates a new list, and replaces the old one in a write-lock
+    pub(crate) fn update(&mut self) -> Result<(), String> {
+        let read_lock = self.inner.read();
+        let (new_allow, new_block) =
+            AllowBlockListInner::load_allow_block_lists(&self.allow_path, &self.block_path)?;
+        let allow_delta = new_allow != read_lock.allow_list;
+        let block_delta = new_block != read_lock.block_list;
+        if allow_delta || block_delta {
+            // Ideally this scope would be atomic
+            let mut mut_inner = {
+                drop(read_lock);
+                self.inner.write()
+            };
+
+            if allow_delta {
+                mut_inner.allow_list = new_allow;
+            }
+            if block_delta {
+                mut_inner.block_list = new_block;
+            }
+        }
+        Ok(())
+    }
+
+    #[cfg_attr(test, allow(unreachable_code, unused_variables))]
+    pub(crate) fn is_ip_allowed(&self, remote_addr: &SocketAddr) -> Result<(), String> {
+        #[cfg(test)]
+        return Ok(());
+
+        let ip = normalize_ip(remote_addr.ip());
+        // whether the peer IP address is blacklisted
+        let read = self.inner.read();
+        if let Some(ip_list) = &read.block_list && ip_list.contains(&ip) {
+            massa_trace!("bootstrap.lib.run.select.accept.refuse_blacklisted", {"remote_addr": remote_addr});
+            Err(format!("IP {} is blacklisted", &ip))
```

Comment:
Should create specific Error instead of String

Reply:
The reason being is that it would be a `BootstrapError::GeneralError(String)` every time.

How about we `struct BootStrapGenErrorMsg(String);` and pass that up?

Reply:
Not possible to replace GeneralError by BlacklistError/WhitelistError ?

Reply:
Even if this falls under the scope of not being a `GeneralError`, we would need to add another variant, which I feel is out of scope for this specific PR

Reply:
I would actually like to do something with the Err enums more broadly. It makes sense in many places to return the inner-value, and not an enum instance, but that is problematic with a lot of inner values being a `String`, and thus not conveying much meaning at the type-level.

I had a quick go at changing `String` to `GenErrString` or similar, and it quickly became apparant that it's a bit out of scope (involves the client side a lot, needs a little bit of design thought put into it, etc).

Would you be happy to put this into a follow-up issue?

Reply:
i think bootstrap error module should return globally `BootstrapError` which is an enum and the load function of file should return for exemple `BootstrapError::InitListError(String)`with the error detailed string in it.

Reply:
@Ben-PH 

I have make an example of what i mean here : https://github.com/massalabs/massa/commit/2cb1de8d7a3a9ec45a1d296840754175dbbe81b2

Reply:
and this for is_ip_allowed : https://github.com/massalabs/massa/commit/8167af85e95e0d2bb8e27c0591d35141f3ba3adf

Reply:
Definitely agree that these are improvements, and the review comments provided. Thanks heaps for going ahead and doing it as well.

I feel there could be improvements to the overall design of the error type system here, but worrying about that is probably an nonconstructive distraction.

Go ahead and bring them in.

Reply:
thanks for adding those changes. I've Boxed the returns to satisfy clippy, I've also made #3608 as a follow-up